File: golang/2816.double-a-number-represented-as-a-linked-list.go

func doubleIt(head *ListNode) *ListNode {
	if head == nil {
		return nil
	}

	head = &ListNode{0, head}

	for cur, next := head, head.Next; next != nil; cur, next = next, next.Next {
		next.Val *= 2
		cur.Val += next.Val / 10
		next.Val %= 10
	}

	if head.Val == 0 {
		return head.Next
	}

	return head
}

File: python/2816.double-a-number-represented-as-a-linked-list.py
# Definition for singly-linked list.
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def doubleIt(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head:
            return None

        cur = head

        if head.val > 4:
            head = ListNode(1, head)

        while cur.next:
            cur.val = (cur.val * 2 + cur.next.val // 5) % 10
            cur = cur.next

        cur.val = (cur.val * 2) % 10

        return head

